MetaMetrics: add 'number_of_nfts' user trait

digiwand opened the pull request:
## Explanation
#13477 Task:
> Add new user trait for 'Number of NFTs'

New user trait: `number_of_nfts`

This involves a slight redesign to the code for `number_of_nft_collections`

## More information

Progresses #13477
Relates to #https://github.com/MetaMask/metamask-extension/pull/14377

digiwand committed:
MetaMetrics: add number_of_nfts trait

digiwand committed:
MetaMetrics: memoize _getAllNFTAddresses

digiwand committed:
MetaMetrics: fix lint / update JSDoc comments

brad-decker commented on app/scripts/controllers/metametrics.js:
```diff
@@ -603,22 +613,30 @@ export default class MetaMetricsController {
+    const allAddresses = this._getAllNFTAddresses(allCollectibles);
+    const uniqueAddresses = new Set(allAddresses);
+    return uniqueAddresses.size;
   }
```
I don't think this is correct...A user may have multiple nfts from the same collection (address) @adonesky1 is that correct?

adonesky1 replied:
Happy to jump on a call @digiwand and clarify these designations if that would help!

digiwand replied:
Makes sense. I think this code is still correct. Maybe there is confusion because updates to both `number_of_nft_collections` and `number_of_nfts` are made in this PR. When I created this logic, I was going off of 

> All collectibles has this shape:
allCollectibles: { [key: string (account)]: { [key: string (chainId)]: Collectible[] } };

> This is an object whose keys are accounts and the values are objects where the keys are chainIds. So, for example, in order to get an aggregate number of NFTs imported you'll have to go through each chainId in each account in this object and sum the length of the Collectibles array.

> In order to get the number of collections you'll need to do the same sort of thing but you'll want the number of unique 'address' properties on the entries. Each collectible has an address property that indicates which collection it is a part of.

from the description in #13477. What you two mentioned sounds aligned with this description, and I think the code has been designed for this. 

This particular method called out here is used to compute the value for `number_of_nft_collections`, while `number_of_nfts` is using 
```
this._getAllNFTAddresses(
    metamaskState.allCollectibles,
).length,
```

After mentioning this, do you two still think we need to update the code? Maybe I'm missing something. Thank you! I also love hopping on calls with you two to pick your brains! 😁  

@brad-decker @adonesky1

brad-decker replied:
Sorry @digiwand -- i misread the code changes. So this would increment for each address regardless of whether it was unique. I conflated the changes with collections here.

adonesky1 replied:
Ah ok. Yes this is correct. Though I think the method _getAllNFTAddresses is a bit of a confusing abstraction even if it is technically working to help get these two traits. I think maybe instead this could be called _getAllNFTs and it could be an array of tuples [address, tokenId], and then _getAllUniqueNFTAddressesLength would filter it down to an array of unique addresses in this array of tuples. Does that make sense?

digiwand replied:
@brad-decker np! easy mix-up. I think it could have helped if I mentioned this in the description too. Updated the description 👍  

@adonesky1 I like this! Thanks. Update the code with your suggestion here: https://github.com/MetaMask/metamask-extension/pull/14495/commits/57b2bc62fe8891bcb753b1c5aeae35b90188dcc2

digiwand committed:
MetaMetrics: update nft trait logic
- https://github.com/MetaMask/metamask-extension/pull/14495#discussion_r855450898

adonesky1 commented on app/scripts/controllers/metametrics.js:
```diff
@@ -603,22 +612,31 @@ export default class MetaMetricsController {
+  _getAllNFTs = memoize((allCollectibles = {}) => {
+    return Object.values(allCollectibles)
+      .flatMap((chainNFTs) => Object.values(chainNFTs))
+      .flat();
```
Is `flat` a `lodash` method? I'm seeing `flatten` in their docs?

adonesky1 replied:
Oh it's a native `array` method nvm!

adonesky1 reviewed (dismissed):
Great work! Left one more nit comment. But not blocking

digiwand committed:
MetaMetrics: rn _getAllNFTs > _getAllNFTsFlattened

adonesky1 reviewed (approved):
LGTM! Great Work!

digiwand commented on app/scripts/controllers/metametrics.js:
```diff
@@ -603,22 +612,31 @@ export default class MetaMetricsController {
-    if (!allCollectibles) {
-      return 0;
-    }
+  _getAllNFTs = memoize((allCollectibles = {}) => {
```
Thanks for hopping on a call with me to explain this! Updated the name to `_getAllNFTsFlattened`. Let me know if you feel like `_getAllNFTsFlat` or another name would suit this better. 

For documentation purposes: 
We had a discussion to deprecate `collection` in favor of `nft` in Slack: https://consensys.slack.com/archives/GTQAGKY5V/p1649274404848719

Which means, `allCollectibles` will be converted to `allNFTs` in the future. Changing `_getAllNFTs` to `_getAllNFTsFlattened` or `_getAllNFTsFlat` will help distinguish the return value from the `allCollectibles`/`allNFTs` state name.
